subsys/nfc/t4t/isodep.c

```diff
@@ -0,0 +1,879 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <kernel.h>
+#include <zephyr/types.h>
+#include <atomic.h>
+#include <nfc/t4t/isodep.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(nfc_t4t_isodep, CONFIG_NFC_T4T_ISODEP_LOG_LEVEL);
+
+#define ISODEP_BLOCK_FIXED_BIT BIT(1)
+#define ISODEP_CRC_LENGTH 2
+
+#define ISODEP_I_BLOCK 0x02
+#define ISODEP_R_BLOCK 0xA2
+#define ISODEP_S_BLOCK 0xC2
+
+#define ISODEP_S_BLOCK_INDICATOR 0x40
+#define ISODEP_R_BLOCK_INDICATOR 0x80
+
+#define ISODEP_BLOCK_NUM_MASK 0x01
+#define ISODEP_MAX_RETRANSMIT_CNT 2
+#define ISODEP_FWT_DEACTIVATION 71680
+#define ISODEP_WTXM_MASK 0x3F
+#define ISODEP_WTXM_MIN 1
+#define ISODEP_WTXM_MAX 59
+
+#define I_BLOCK_DID_BIT BIT(3)
+#define I_BLOCK_NAD_BIT BIT(2)
+#define I_BLOCK_CHAINING_BIT BIT(4)
+#define R_BLOCK_NAK BIT(4)
+#define S_BLOCK_WTX_MASK 0x30
+
+#define T4T_FSD_MIN 16
+
+#define T4T_RATS_CMD 0xE0
+#define T4T_RATS_DID_MASK 0x0F
+#define T4T_RATS_FSDI_MASK 0xF0
+#define T4T_RATS_FSDI BIT(3)
+#define T4T_RATS_CMD_LEN 0x02
+
+#define T4T_FWT_ACTIVATION 71680
+#define T4T_FWT_DELTA 49152
+
+#define NFCA_T4T_FWT_T_FC (164 * 1356UL)
+#define T4T_FC_IN_MS 13560UL
+
+#define T4T_FWT_TO_MS(_fwt)                   \
+	ceiling_fraction((_fwt), T4T_FC_IN_MS)
+
+#define T4T_RATS_FDT (T4T_FWT_ACTIVATION +        \
+		T4T_FWT_DELTA + NFCA_T4T_FWT_T_FC)
+
+#define T4T_DID_MAX 14
+
+#define T4T_ATS_MAX_LEN 20
+#define T4T_ATS_MIN_LEN 2
+#define T4T_ATS_T0_TA_PRESENT BIT(4)
+#define T4T_ATS_T0_TB_PRESENT BIT(5)
+#define T4T_ATS_T0_TC_PRESENT BIT(6)
+#define T4T_ATS_TA_EQUAL_DIVISOR BIT(7)
+#define T4T_ATS_T0_FSCI_MASK 0x0F
+#define T4T_ATS_TA_DIVISOR_PL 0x07
+#define T4T_ATS_TA_DIVISOR_PL_OFFSET 1
+#define T4T_ATS_TA_DIVISOR_LP_OFFSET 3
+#define T4T_ATS_TA_DIVISOR_LP 0x70
+#define T4T_ATS_TB_FWI_DEFAULT 0x04
+#define T4T_ATS_TB_FWI_MASK 0xF0
+#define T4T_ATS_TB_FWI_OFFSET 4
+#define T4T_ATS_TB_SFGI_MASK 0x0F
+#define T4T_ATS_TC_DID BIT(1)
+#define T4T_ATS_TC_NAD BIT(0)
+#define T4T_ATS_SFGT_DEFAULT 6780
+
+/* Convert FWI to FWT. NFC Forum Digital Specification 2.0 14.8.1. */
+#define T4T_FWI_TO_FWT(_fwi) (256u * 16u * (1 << (_fwi)))
+
+/* Calculate SFGT margin. NFC Forum Digital Specification 2.0 14.8.2. */
+#define T4T_SFGT_DELTA(_sfgi) (384 * (1 << (_sfgi)))
```

Comment:
why `384` while line 82 has `256u`?

Reply:
Formula said that: NFC Forum Digital Specification 2.0 14.8.2 Value comes from it